Custom expression: fix unary parsing with boolean

ariya opened the pull request:
Fixes #22216

To verify:
1. New, SQL query
2. Type `select true as awesome, false as terrible`, run and save it as `TestBool`
3. Explore results
4. Filter, Custom expression and type `[AWESOME] = NOT [TERRIBLE]`

(Another example that demonstrates the issue: use more unary operators, `NOT NOT`)

### Before this PR

Can't click the `Done` button (it's disabled) since the expression is _incorrectly_ rejected.

![image](https://user-images.githubusercontent.com/7288/166728653-183a24ef-44c2-4c49-b195-a98c12b9e46c.png)

From the browser console:
```
compile error Error: Unexpected operator NOT
```

### After this PR

The expression is correctly accepted as valid.

![image](https://user-images.githubusercontent.com/7288/166728548-0188fddb-1234-4cee-a99a-23ea4da911a5.png)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/22429)
<!-- Reviewable:end -->

Commit:
Custom expression: fix unary parsing with boolean

ariya commented on frontend/src/metabase/lib/expressions/recursive-parser.js:
```diff
@@ -332,7 +333,11 @@ export const adjustBooleans = tree =>
+            if (
+              !Array.isArray(operand) ||
+              args[index] !== "boolean" ||
+              operator === "not"
```
This exception is necessary so that `NOT X` (where `X` is a boolean field) isn't formatted as `NOT X = True` (redundant, and very confusing).

akiselev replied:
I think we should wait for the backend PR on this one so we can remove this exception entirely. I added that block to rewrite calls like `case([boolean], 0, 1)` to `case([boolean] = True, 0, 1)` because of the same backend error:

![image](https://user-images.githubusercontent.com/653604/166784329-0d5758fb-ca28-443d-b0c8-a4a619814517.png)


![image](https://user-images.githubusercontent.com/653604/166782966-3ecff072-8407-47a7-b57a-553be144b9b1.png)

With your exception added, it makes the functionality inconsistent - `case(NOT [boolean], ...)` doesnt work but `case([boolean], ...)` does:


![image](https://user-images.githubusercontent.com/653604/166794246-aefdc27e-4031-4555-a334-a98caf3bb62d.png)


 We might as well remove it as it wasn't a particularly well thought out piece of code, especially since not all SQL servers support booleans (Oracle...). It should be behind a driver feature and/or have a normalizer on the backend

Commit:
Fix matching booleans with expression

ariya commented on frontend/src/metabase/lib/expressions/resolver.js:
```diff
@@ -166,12 +166,13 @@ export function resolve(expression, type = "expression", fn = undefined) {
+  }
+
+  if (type === "expression" && typeof expression === "boolean") {
+    return expression;
```
This is rather a quick hack right now to let `F(X) < Y` marked as invalid (where F is a function returning boolean).
Expect a more comprehensive follow-up PR to tackle extensive comparison ops with booleans.

Commit:
Fix E2E assertion

akiselev reviewed (changes requested):
I think we should wait for the backend PR on this one so we can remove that boolean exception entirely. I added that block to rewrite calls like `case([boolean], 0, 1)` to `case([boolean] = True, 0, 1)` because of the same backend error:

![image](https://user-images.githubusercontent.com/653604/166784329-0d5758fb-ca28-443d-b0c8-a4a619814517.png)


![image](https://user-images.githubusercontent.com/653604/166782966-3ecff072-8407-47a7-b57a-553be144b9b1.png)

With your exception added, it makes the functionality inconsistent - `case(NOT [boolean], ...)` doesnt work but `case([boolean], ...)` does:


![image](https://user-images.githubusercontent.com/653604/166794246-aefdc27e-4031-4555-a334-a98caf3bb62d.png)


 We might as well remove it as it wasn't a particularly well thought out piece of code, especially since not all SQL servers support booleans (Oracle...). It should be behind a driver feature and/or have a normalizer on the backend

escherize committed:
make unary not work, by transforming it to using = ... false

- values ["not" x] to ["=" x false]

Commit:
Remove boolean expression rewrite on the front-end

ariya commented:
> I think we should wait for the backend PR on this one so we can remove that boolean exception entirely. I added that block to rewrite calls like `case([boolean], 0, 1)` to `case([boolean] = True, 0, 1)` because of the same backend error:

Agree. I removed `adjustBooleans` for the time being. While tracking this down with @escherize, we discovered a very much related issue with unary operator anymore (not unique to `NOT`, _pun intended_): #22455. Hopefully if that issue is sorted out, it's easier on the front-end to tackle various Boolean-related challenges.

escherize committed:
allow using fields as filters in and or and not clauses

escherize committed:
Merge branch 'fix-logical-unary' of github.com:metabase/metabase into fix-logical-unary
